### DISCUSSION TITLE
HOWTO using I2C multi display (hardware mode).

### DISCUSSION THREAD

**Discussion by TrDA-hab** [arendst/Tasmota#13166] [Show and tell]
## 0. About using I2C multi display.   
 - [How to use I2C bus](https://github.com/arendst/Tasmota/discussions/10827).  
 - Adafruit info [about using 74HC138](https://learn.adafruit.com/delorean-time-circuit/circuit-trickery).  
 - 74hc138 datasheet [1](https://static.chipdip.ru/lib/935/DOC011935338.pdf) , [2](https://www.diodes.com/assets/Datasheets/74HC138.pdf) (High-Speed CMOS Logic 3- to 8- Line Demultiplexer).   
 - Tasmota [commands](https://tasmota.github.io/docs/Commands/#displays) for controlling displays.  
 - Tasmota [examples](https://tasmota.github.io/docs/Displays/#rule-examples-for-scripting-examples-see-scripting-docs) for displaying information on the display.  
 - Tasmota [use Buttons](https://tasmota.github.io/docs/Buttons-and-Switches/#button).   
 - Tasmota [use Rules](https://tasmota.github.io/docs/Rules/).    
 - You can use any I2C display (only if supported by Tasmota). SPI displays have not been tested, but should work too.
 - You can connect x3 I2C displays.  
   - or up to x6 maximum, if you modify the circuit (if you are using ESP8266).  
   - or up to x8 maximum, if you modify the circuit (if you are using ESP32).   
 - You can only send information to one display currently active (including the corresponding relay).  
 - The display backlight can only be controlled on the active display. The inactive display saves the display subgrid state.  
 - **!!!Caution!!!** 
   - Displays must be initialized every time the ESP8266/ESP32 is powered off/on.  
   - After Hardware or Software Reset, all display settings are saved.**  

## Other projects where multi display is used:
 - Tasmota [SPI multi display (hardware mode)](https://github.com/arendst/Tasmota/discussions/13161).  
 - Tasmota [SPI+I2C multi display (software mode)](https://github.com/arendst/Tasmota/discussions/13222).  
 - [ESP32 and 8 OLEDs arranged in a circle](https://youtu.be/KZTIBoHtouM). 
 - [Multiple displays on the I2C bus](https://www.youtube.com/watch?v=E9FTQyBYwAE).  
 - [8 Multiplexed OLED /w ESP32 and U8g2](https://www.youtube.com/watch?v=aMgIxXwtHbw).  
 - [SPI OLED multiple displays with an Arduino](https://youtu.be/YCkFFtVEEG4).  
 - [Multiple SPI based Displays with TFT_eSPI on an esp32](https://youtu.be/cCgNHIHijhs).  
 - [Multiple Displays with Arduino](https://youtu.be/yef23sJjiU0).  
 - [Multiple OLED screens](https://youtu.be/TOMkXJWdB4w).  
 - [Multiple OLED Displays / Multiplex TC9548](https://youtu.be/Y9OyLMUgoFk).   

## 1. I2C multi display (minimal cofiguration).  
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/I2C%20multi%20display/4161.jpg)  
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/I2C%20multi%20display/4162.jpg)   

Wemos Pin|GPIO|Component|I2C display|Сomment|
:-:|:-:|:-:|:-:|:-:
D1|05|I2C SCL|SCL|-
D2|12|I2C SDA|SDA|-
D3|00|Relay_1|-|On/Off display#1
D4|02|Relay_2|-|On/Off display#2
D5|14|Relay_3|-|On/Off display#3
N/C|-|Relay_4|backligh|On/Off backlight

## How to use it.  
 - You must add support for display in `my_user_config.h` file, or flash our ESP8266 module with `tasmota-display.bin` file.   
 - **Caution!!! Displays must be initialized every time the ESP8266 is powered off/on.**
 - To initialize the displays, run the commands in the console:  
   `Backlog Power3 1; Power4 1; Restart 1`  // To initialize display#3 and restart Tasmota.  
   `Backlog Power2 1; Power4 1; Restart 1`  // To initialize display#2 and restart Tasmota.    
   `Backlog Power1 1; Power4 1; Restart 1`  // To initialize display#1 and restart Tasmota.    
 - Run the command in the console  to run the "Interlock" mode, for the possibility of individual output of information on displays (optional):  
   `SetOption73 ON`  // Enable detach buttons from relays.  
   `Interlock 1,2,3` // Group Relay1 and Relay2 and Relay3 in "group 1" (only one display can work at a time).  
   `Interlock 1`     // Enable relay interlock mode.  
 - Run the command in the console to modify the interface (optional):   
   `Backlog WebButton1 LCD#1; WebButton2 LCD#2; WebButton3 LCD#3; WebButton4 Light`  
 - Run the commands in the console to test displays:  
   `Backlog Power4 0; Power1 1; Power4 1` // For run display#1.  
   `Backlog Power4 0; Power2 1; Power4 1` // For run display#2.  
   `Backlog Power4 0; Power3 1; Power4 1` // For run display#3.  

## 2. I2C multi display (maximal cofiguration).   
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/I2C%20multi%20display/4171.jpg)   
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/I2C%20multi%20display/4172.jpg)    

Wemos Pin|GPIO|Component|I2C display|Сomment|
:-:|:-:|:-:|:-:|:-:
D1|05|I2C SCL|SCL|-
D2|12|I2C SDA|SDA|-
D3|00|Relay_1|-|On/Off display#1
D4|02|Relay_2|-|On/Off display#2
D5|14|Relay_3|-|On/Off display#3
N/C|-|Relay_4|backligh|On/Off backlight
TX|01|Button_1|-|Switching between displays
RX|03|Button_2|backligh|On/Off backlight  

## How to use it.  
 - You must add support for display in `my_user_config.h` file, or flash our ESP8266 module with `tasmota-display.bin` file.   
 - **Caution!!! Displays must be initialized every time the ESP8266 is powered off/on.**
 - To initialize the displays, run the commands in the console:  
   `Backlog Power3 1; Power4 1; Restart 1`  // To initialize display#3 and restart Tasmota.  
   `Backlog Power2 1; Power4 1; Restart 1`  // To initialize display#2 and restart Tasmota.  
   `Backlog Power1 1; Power4 1; Restart 1`  // To initialize display#3 and restart Tasmota.  
 - Run the command in the console  to run the "Interlock" mode, for the possibility of individual output of information on displays (optional):  
   `SetOption73 ON`  // Enable detach buttons from relays.  
   `Interlock 1,2,3` // Group Relay1 and Relay2 and Relay3 in "group 1" (only one display can work at a time).  
   `Interlock 1`     // Enable relay interlock mode.  
 - Run the command in the console to modify the interface (optional):   
   `Backlog WebButton1 LCD#1; WebButton2 LCD#2; WebButton3 LCD#3; WebButton4 Light`  
 - Run the commands in the console to test displays:  
   `Backlog Power4 0; Power1 1; Power4 1` // For run display#1.  
   `Backlog Power4 0; Power2 1; Power4 1` // For run display#2.  
   `Backlog Power4 0; Power3 1; Power4 1` // For run display#3.  
 - Assigns button#1 to Switching between displays (1x press up, 2x press down, 3x press down):    
   `Rule1`   
   `ON button1#state=10 DO Backlog Power4 0; Power1 1; Power4 1 ENDON`  // 1x press up to run display#1.      
   `ON button1#state=11 DO Backlog Power4 0; Power2 1; Power4 1 ENDON`  // 3x press up to run display#2.      
   `ON button1#state=12 DO Backlog Power4 0; Power3 1; Power4 1 ENDON`  // 3x press up to run display#5.      
   `Rule1 1`  // Enable Rule1.  
 - Assigns button#2 to On/Off Display backlight (1x press up, 2x press down):   
   `Rule2`   
   `ON button2#state=10 DO Power4 0 ENDON`  // 1x press up to OFF Display backlight.  
   `ON button2#state=11 DO Power4 1 ENDON`  // 2x press up to ON Display backlight.  
   `Rule2 1` // Enable Rule2.  

### Video of the display`s work:   
 - [Video#1](https://youtu.be/zgQKJD9gmQw).   
 - [Video#2](https://youtu.be/Cm0D1HlTeSg).  

![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/I2C%20multi%20display/20210919_193223.jpg) 
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/I2C%20multi%20display/20210918_152007.jpg)  
![](https://raw.githubusercontent.com/TrDA-hab/Projects/master/I2C%20multi%20display/20210919_192536.jpg)  

**Best regards,   
TrDA.**

**Comment by b-acksla-sh** [comment]
Thanks! I was able to get three OLEDs working with a single ESP8266 using this guide. 

I used SH1106 OLEDs which didn't need that initialization sequence. I also used a 3-to-8 74HC4051 analog multiplexer instead since those modules are more easily available in my part of the world. The unused A2 input needed to be grounded otherwise the displays would show glitched graphics.

I didn't have any spare GPIOs so I skipped the virtual gpio and the relay interlocking and resorted to just plain old binary:

Select Display 1: `Power1 0; Power2 0`
Select Display 2: `Power1 0; Power2 1`
Select Display 3: `Power1 1; Power2 0`

https://www.youtube.com/watch?v=HmSjiquPmQo
